WalletWasabi/Affiliation/AffiliateInformationUpdater.cs

```diff
@@ -0,0 +1,297 @@
+		Clients = clients;
+		Signer = signer;
+
+		CoinjoinRequests = new();
+		RoundData = new();
+		RoundsToUpdate = new();
+		RoundsToRemove = new();
+
+		AddHandlers(arena);
+	}
+
+	private Arena Arena { get; }
+	private Signer Signer { get; }
+	private ImmutableDictionary<AffiliationFlag, AffiliateServerHttpApiClient> Clients { get; }
+	private Dictionary<uint256, Dictionary<AffiliationFlag, byte[]>> CoinjoinRequests { get; }
+	private ConcurrentDictionary<uint256, RoundData> RoundData { get; }
+	private Queue<uint256> RoundsToUpdate { get; }
+	private Queue<uint256> RoundsToRemove { get; }
+
+	public override void Dispose()
+	{
+		RemoveHandlers(Arena);
+	}
+
+	public IReadOnlyDictionary<uint256, IReadOnlyDictionary<AffiliationFlag, byte[]>> GetCoinjoinRequests()
+	{
+		return (IReadOnlyDictionary<uint256, IReadOnlyDictionary<AffiliationFlag, byte[]>>)CoinjoinRequests.ToDictionary(x => x.Key, x => (IReadOnlyDictionary<AffiliationFlag, byte[]>)x.Value);
+	}
+
+	protected override async Task ActionAsync(CancellationToken cancel)
+	{
+		await UpdateCoinjoinRequestsAsync();
+		RemoveRounds();
+	}
+
+	private async Task UpdateCoinjoinRequestsAsync(uint256 roundId, RoundData roundData, AffiliationFlag affiliationFlag, AffiliateServerHttpApiClient affiliateServerHttpApiClient)
+	{
+		Body body = roundData.GetAffiliationData(affiliationFlag);
+		byte[]? result = await GetCoinjoinRequestAsync(affiliateServerHttpApiClient, body);
+
+		if (result is not null)
+		{
+			if (!CoinjoinRequests.ContainsKey(roundId))
+			{
+				CoinjoinRequests.Add(roundId, new());
+			}
+
+			Dictionary<AffiliationFlag, byte[]> coinjoinRequests = CoinjoinRequests[roundId];
+
+			if (coinjoinRequests.ContainsKey(affiliationFlag))
+			{
+				throw new InvalidOperationException("The coinjoin request is already set.");
+			}
+
+			coinjoinRequests.Add(affiliationFlag, result);
+		}
+		else
+		{
+			Logging.Logger.LogWarning($"Cannot get coinjoin request from affiliate server '{affiliationFlag}'.");
+		}
+	}
+
+	private async Task UpdateCoinjoinRequestsAsync(uint256 roundId)
+	{
+		if (!RoundData.TryGetValue(roundId, out RoundData? roundData))
+		{
+			throw new InvalidOperationException($"The round ({roundId}) does not exist.");
+		}
+
+		roundData.Lock();
+
+		await Task.WhenAll(Clients.Select(x => UpdateCoinjoinRequestsAsync(roundId, roundData, x.Key, x.Value)));
+	}
+
+	private async Task UpdateCoinjoinRequestsAsync()
+	{
+		while (RoundsToUpdate.Any())
+		{
+			try
+			{
+				uint256 roundId = RoundsToUpdate.Dequeue();
+				if (!RoundsToRemove.Contains(roundId))
+				{
+					await UpdateCoinjoinRequestsAsync(roundId);
+				}
+			}
+			catch (Exception exception)
+			{
+				Logging.Logger.LogError(exception.Message);
+			}
+		}
+	}
+
+	private void RemoveRound(uint256 roundId)
+	{
+		if (!RoundData.TryRemove(roundId, out var roundData))
+		{
+			throw new InvalidOperationException($"The round ({roundId}) does not exist.");
+		}
+
+		if (CoinjoinRequests.ContainsKey(roundId))
```

Comment:
Avoid tester-doer pattern and use TryRemove

Reply:
As far as I know, `Dictionary` doesn't implement `TryRemove`, see [this](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2?view=net-7.0).

Reply:
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/76e434c939215dbff81a865ba3091f1afafa2ec0.